File: app/src/main/java/com/billpayments/core/cache/Freshness.kt
package com.billpayments.core.cache

import com.billpayments.core.time.Clock
import java.time.Duration
import java.time.Instant

class Freshness(private val clock: Clock, private val maxAge: Duration) {
    private var fetchedAt: Instant? = null

    fun markFetched() {
        fetchedAt = clock.now()
    }

    fun isStale(): Boolean {
        val fetched = fetchedAt ?: return true
        return Duration.between(fetched, clock.now()) >= maxAge
    }
}

File: BillPayments/Core/Freshness.swift
import Foundation

final class Freshness {
    private let clock: Clock
    private let maxAge: TimeInterval
    private var fetchedAt: Date?

    init(clock: Clock, maxAge: TimeInterval) {
        self.clock = clock
        self.maxAge = maxAge
    }

    func markFetched() { fetchedAt = clock.now() }

    func isStale() -> Bool {
        guard let fetchedAt else { return true }
        return clock.now().timeIntervalSince(fetchedAt) >= maxAge
    }
}
